src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -880,22 +890,79 @@ QgsLegendModel::QgsLegendModel( QgsLayerTree *rootNode, QObject *parent )
 QVariant QgsLegendModel::data( const QModelIndex &index, int role ) const
 {
   // handle custom layer node labels
-  if ( QgsLayerTreeNode *node = index2node( index ) )
+
+  QgsLayerTreeNode *node = index2node( index );
+  QgsLayerTreeModelLegendNode *ltmln = index2legendNode( index ); // Possibly useless
+  QgsLayerTreeLayer *nodeLayer = QgsLayerTree::isLayer( node ) ? QgsLayerTree::toLayer( node ) : nullptr;
+  if ( nodeLayer && ( role == Qt::DisplayRole || role == Qt::EditRole ) )
   {
-    if ( QgsLayerTree::isLayer( node ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) && !node->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
+    QString name;
+    QgsSymbolLegendNode *symnode = nullptr;
+    QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
+    if ( ltmln )
+    {
+      symnode = qobject_cast<QgsSymbolLegendNode *>( ltmln );
+      name = symnode->symbolLabel();
+    }
+    else
     {
-      QgsLayerTreeLayer *nodeLayer = QgsLayerTree::toLayer( node );
-      QString name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
-      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() && role == Qt::DisplayRole )
+      //finding the first label that is stored
+      name = nodeLayer->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = nodeLayer->name();
+      if ( name.isEmpty() )
+        name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = node->name();
+      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() )
       {
-        QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
         if ( vlayer && vlayer->featureCount() >= 0 )
+        {
           name += QStringLiteral( " [%1]" ).arg( vlayer->featureCount() );
+          Q_UNUSED( ltmln );
+          Q_UNUSED( symnode );
+          return name;
+        }
       }
-      return name;
     }
-  }
 
+    bool evaluate = vlayer ? !nodeLayer->labelExpression().isEmpty() : false;
+
+    if ( evaluate || name.contains( "[%" ) )
+    {
+      QgsExpressionContext expressionContext;
+      if ( vlayer )
+        connect( vlayer, &QgsVectorLayer::symbolFeatureCountMapChanged, this, &QgsLegendModel::forceRefresh );
+
+      if ( mLayoutLegend )
+        expressionContext = mLayoutLegend->createExpressionContext();
+      else if ( vlayer )
+        expressionContext = vlayer->createExpressionContext();
+      else
+        expressionContext = QgsExpressionContext();
```

Comment:
This is redundant with default initialization on https://github.com/qgis/QGIS/pull/9648/files#diff-8d62b51ea1de0f0b40b8e73b05037a19R933

Reply:
true, I'll remove this one